File: BRReusableQueue/BRReusableQueue.m
//
//  BRReusableQueue.m
//  BRReusableQueue
//
//  Created by Cornelius Horstmann on 07.02.16.
//  Copyright © 2016 Cornelius Horstmann. All rights reserved.
//

#import "BRReusableQueue.h"
#import <UIKit/UIKit.h>

@implementation BRReusableQueue {
    NSCache *_reusables;
}

+ (instancetype)sharedQueue {
    static dispatch_once_t onceToken;
    static BRReusableQueue *sharedQueue;
    dispatch_once(&onceToken, ^{
        sharedQueue = [[BRReusableQueue alloc] init];
    });
    return sharedQueue;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)init {
    self = [super init];
    if (self ) {
        _reusables = [NSCache new];
        self.emptyOnMemoryWarningNotification = YES;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didReceiveMemoryWarning:) name:UIApplicationDidReceiveMemoryWarningNotification object:nil];
    }
    return self;
}

- (void)enqueueReusable:(id <BRReusable>)reusable {
    NSString *identifier = [self identifierFromReusableClass:[reusable class]];
    NSMutableArray *objects = [[_reusables objectForKey:identifier] mutableCopy];
    if (!objects) {
        objects = [NSMutableArray new];
    }
    if ([reusable respondsToSelector:@selector(prepareForReuse)]) {
        [reusable performSelector:@selector(prepareForReuse)];
    }
    [objects addObject:reusable];
    [_reusables setObject:objects forKey:identifier];
}

- (id <BRReusable>)dequeueReusableWithIdentifier:(NSString *)identifier {
    NSMutableArray *objects = [_reusables objectForKey:identifier];
    if (!objects) {
        return nil;
    }
    id<BRReusable> object = [objects lastObject];
    if (!object) {
        return nil;
    }
    [objects removeObject:object];
    [_reusables setObject:objects forKey:identifier];
    return object;
}

- (id <BRReusable>)dequeueReusableWithClass:(Class)reuseClass {
    return [self dequeueReusableWithIdentifier:[self identifierFromReusableClass:reuseClass]];
}

- (id <BRReusable>)dequeueOrCreateReusableWithClass:(Class)reusableClass {
    id <BRReusable> reusable = [self dequeueReusableWithClass:reusableClass];
    if (reusable) {
        return reusable;
    }
    if (![reusableClass respondsToSelector:@selector(newForReuse)]) {
        return nil;
    }
    reusable = [reusableClass performSelector:@selector(newForReuse)];
    return reusable;
}

- (void)emptyQueue {
    [_reusables removeAllObjects];
}

#pragma mark private

- (NSString *)identifierFromReusableClass:(Class)reusableClass {
    if ([reusableClass respondsToSelector:@selector(reuseIdentifier)]) {
        return [reusableClass performSelector:@selector(reuseIdentifier)];
    }
    return NSStringFromClass(reusableClass);
}

- (void)didReceiveMemoryWarning:(NSNotification *)notification {
    if (!self.emptyOnMemoryWarningNotification) {
        return;
    }
    [self emptyQueue];
}

@end

File: BRReusableQueue/BRReusableQueue.swift
//
//  BRReusableQueue.swift
//  BRReusableQueue
//
//  Created by Cornelius Horstmann on 30.01.16.
//  Copyright © 2016 Cornelius Horstmann. All rights reserved.
//

import Foundation
import UIKit

@objc public protocol BRReusable: class {
    optional static func reuseIdentifier() -> String // if not implemented, classname is used
    optional static func newForReuse() -> AnyObject? // if implemented, dequeueOrCreateReusableWithClass can be used
    optional func prepareForReuse() -> Void
}

@objc public class BRReusableQueue : NSObject {
    var reusables: NSCache = NSCache()
    public var emptyOnMemoryWarningNotification = true
    
    public static let sharedQueue = BRReusableQueue()
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    override init () {
        super.init()
        NSNotificationCenter.defaultCenter().addObserver(
            self,
            selector: "didReceiveMemoryWarning",
            name: UIApplicationDidReceiveMemoryWarningNotification,
            object: nil)
    }
    
    public func enqueueReusable(reusable: BRReusable) -> Void {
        let identifier = identifierFromReusableClass(reusable.dynamicType)
        var objects: Array<BRReusable> = reusables.objectForKey(identifier) as? Array<BRReusable> ??  Array<BRReusable>()
        reusable.prepareForReuse?()
        objects.append(reusable)
        reusables.setObject(objects, forKey: identifier)
    }
    
    public func dequeueReusableWithIdentifier(identifier: String) -> BRReusable? {
        if var objects: Array<BRReusable> = reusables.objectForKey(identifier) as? Array<BRReusable> where
            objects.count > 0 {
            let reusable = objects.removeFirst()
            reusables.setObject(objects, forKey: identifier)
            return reusable
        }
        return nil
    }
    
    public func dequeueReusableWithClass(reusableClass: BRReusable.Type) -> BRReusable? {
        return dequeueReusableWithIdentifier(identifierFromReusableClass(reusableClass))
    }
    
    public func dequeueOrCreateReusableWithClass(reusableClass: BRReusable.Type) -> BRReusable? {
        if let reusable = dequeueReusableWithClass(reusableClass) {
            return reusable
        }
        let createdReusable = reusableClass.newForReuse?()
        if let reusable = createdReusable as? BRReusable {
            return reusable
        }
        return nil
    }
    
    public func emptyQueue() -> Void {
        reusables.removeAllObjects()
    }
    
    func didReceiveMemoryWarning() -> Void {
        if !emptyOnMemoryWarningNotification {
            return
        }
        emptyQueue()
    }
    
    private func identifierFromReusableClass(reusable: BRReusable.Type) -> String {
        if let identifier = reusable.reuseIdentifier?() {
            return identifier
        }
        return  "\(reusable)"
    }
    
}